src/CoreWCF.Primitives/src/CoreWCF/Runtime/SimpleAsyncLock.cs

```diff
@@ -0,0 +1,61 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace CoreWCF.Runtime
+{
+    /// <summary>
+    /// This class provides a locking mechanism compatible with task async patterns.
+    /// </summary>
+    /// <remarks>
+    /// Unlike the <see cref="AsyncLock"/> it does not have the complexity of lock
+    /// tracking and it ensures SemaphoreSlim does not leak memory or handles by hiding
+    /// the access to certain members which would create resources that require disposal.
+    /// </remarks>
+    internal sealed class SimpleAsyncLock
```

Comment:
Can you rename this to NonReentrantAsyncLock? The problem that AsyncLock was trying to solve was allowing you to continue to take a lock on the same task. Regular monitor locks tie the lock to the thread, but with async code you can hop threads which is why monitor locks don't work. This code can't be reentered and can deadlock. E.g. something like this will deadlock:
```c#
public async Task DoSomething()
{
    using(_asyncLock.TakeLockAsync())
    {
        await DoSomethingElse();
    }
}
public async Task DoSomethingElse()
{
    using(_asyncLock.TakeLockAsync())
    {
        Console.WriteLine("Doing something else");
    }
}
```
My initial approach in AsyncLock was to track lock taken in an AsyncLocal, but that only works if you have a linear set of Task methods. With that approach if you do something like this:
```c#
public async Task DoSomething()
{
    using(_asyncLock.TakeLockAsync())
    {
        var task1 = DoSomethingElse();
        var task2 = DoSomethingElse();
        await Task.WhenAll(task1, task2);
    }
}
public async Task DoSomethingElse()
{
    using(_asyncLock.TakeLockAsync())
    {
        Console.WriteLine("Doing something else");
    }
}
```
Then both task1 and task2 would be able to enter the lock at the same time, which defeats the purpose of the lock. I have an idea how to address this, but the book keeping code is quite complicated and I haven't had the 2+ days available to "do it right". So for now let's rename this to NonReentrantAsyncLock so anybody introducing new usage is very clear about it's limitations. What might be a nice addition is to add some DEBUG only code which detects reentrance usage and asserts. This would be a static AsyncLocal<bool> that you set to true when acquiring the lock, and set to false when releasing. Before acquiring the lock, check to see if the AsyncLocal is true and assert. The only problem with this approach is it would fire if you are using two different NonReentrantAsyncLock instances on the call path. I'm not requiring that for this PR, basically giving you my stream of thought. The only change required here is to rename the class.

Reply:
Reentrancy is indeed a very strong argument to keep the versions separate. I will rename them both to `NonReentrantAsyncLock` and `ReentrantAsyncLock` to be explicit in both cases. 

This way the reentrant version also can then be tackled separately. I stumbled over the [ReentrantSemaphore of the VSSDK](https://github.com/microsoft/vs-threading/blob/main/src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs). We might use it or adopt it into CoreWCF. It has quite a flexible design on the different reentrancy use cases.